File: backend/internal/auth/twitter.go
package auth

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

type TwitterUser struct {
	ID              string `json:"id"`
	Username        string `json:"username"`
	Name            string `json:"name"`
	ProfileImageURL string `json:"profile_image_url"`
}

func BuildTwitterAuthURL(clientID, callback, state, challenge string) string {
	q := url.Values{}
	q.Set("response_type", "code")
	q.Set("client_id", clientID)
	q.Set("redirect_uri", callback)
	q.Set("scope", "tweet.read users.read offline.access")
	q.Set("state", state)
	q.Set("code_challenge", challenge)
	q.Set("code_challenge_method", "S256")
	return "https://twitter.com/i/oauth2/authorize?" + q.Encode()
}

func ExchangeTwitterCode(clientID, clientSecret, callback, code, verifier string) (accessToken string, err error) {
	form := url.Values{}
	form.Set("code", code)
	form.Set("grant_type", "authorization_code")
	form.Set("client_id", clientID)
	form.Set("redirect_uri", callback)
	form.Set("code_verifier", verifier)

	req, err := http.NewRequest(http.MethodPost, "https://api.twitter.com/2/oauth2/token", strings.NewReader(form.Encode()))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.SetBasicAuth(clientID, clientSecret)

	client := &http.Client{Timeout: 15 * time.Second}
	res, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer res.Body.Close()
	body, _ := io.ReadAll(io.LimitReader(res.Body, 1<<20))
	if res.StatusCode >= 300 {
		return "", fmt.Errorf("twitter token exchange failed (%d): %s", res.StatusCode, truncate(string(body), 200))
	}
	var out struct {
		AccessToken string `json:"access_token"`
	}
	if err := json.Unmarshal(body, &out); err != nil {
		return "", err
	}
	if out.AccessToken == "" {
		return "", fmt.Errorf("empty twitter access token")
	}
	return out.AccessToken, nil
}

func FetchTwitterMe(accessToken string) (*TwitterUser, error) {
	req, err := http.NewRequest(http.MethodGet, "https://api.twitter.com/2/users/me?user.fields=profile_image_url", nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+accessToken)
	client := &http.Client{Timeout: 15 * time.Second}
	res, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer res.Body.Close()
	body, _ := io.ReadAll(io.LimitReader(res.Body, 1<<20))
	if res.StatusCode >= 300 {
		return nil, fmt.Errorf("twitter /me failed (%d): %s", res.StatusCode, truncate(string(body), 200))
	}
	var out struct {
		Data TwitterUser `json:"data"`
	}
	if err := json.Unmarshal(body, &out); err != nil {
		return nil, err
	}
	return &out.Data, nil
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n]
}

File: frontend/src/lib/server/twitter.ts
import { twitterCallbackUrl, twitterClientId, twitterClientSecret } from "./config";

export type TwitterUser = {
  id: string;
  username: string;
  name: string;
  profile_image_url?: string;
};

export function buildTwitterAuthUrl(
  state: string,
  challenge: string,
): string {
  const params = new URLSearchParams({
    response_type: "code",
    client_id: twitterClientId(),
    redirect_uri: twitterCallbackUrl(),
    scope: "tweet.read users.read offline.access",
    state,
    code_challenge: challenge,
    code_challenge_method: "S256",
  });
  return `https://x.com/i/oauth2/authorize?${params.toString()}`;
}

export async function exchangeTwitterCode(
  code: string,
  verifier: string,
): Promise<string> {
  const body = new URLSearchParams({
    code,
    grant_type: "authorization_code",
    client_id: twitterClientId(),
    redirect_uri: twitterCallbackUrl(),
    code_verifier: verifier,
  });

  const res = await fetch("https://api.twitter.com/2/oauth2/token", {
    method: "POST",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
      Authorization: `Basic ${Buffer.from(`${twitterClientId()}:${twitterClientSecret()}`).toString("base64")}`,
    },
    body,
  });

  const json = (await res.json()) as { access_token?: string; error?: string };
  if (!res.ok || !json.access_token) {
    throw new Error(json.error ?? `Twitter token exchange failed (${res.status})`);
  }
  return json.access_token;
}

export async function fetchTwitterMe(
  accessToken: string,
): Promise<TwitterUser> {
  const res = await fetch(
    "https://api.twitter.com/2/users/me?user.fields=profile_image_url",
    {
      headers: { Authorization: `Bearer ${accessToken}` },
    },
  );
  const json = (await res.json()) as { data?: TwitterUser; detail?: string };
  if (!res.ok || !json.data) {
    throw new Error(json.detail ?? `Twitter /me failed (${res.status})`);
  }
  return json.data;
}
